runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -0,0 +1,239 @@
+      on_reload = function(_, b)
+        hint_cache_by_buf[b] = nil
+      end,
+    })
+  end
+  local hints_by_client = bufstate.client_hint
+  local hints_by_lnum = hints_by_client[client_id]
+  local client = vim.lsp.get_client_by_id(client_id)
+
+  -- To reduce flicker, preserve old resolved hints until new ones are resolved.
+
+  local new_hints_by_lnum = vim.defaulttable()
+  local num_unprocessed = #hints
+  if num_unprocessed == 0 then
+    hints_by_client[client_id] = {}
+    bufstate.version = version
+    return
+  end
+
+  for _, hint in ipairs(hints) do
+    local lnum = hint.position.line
+    new_hints_by_lnum[lnum][#new_hints_by_lnum[lnum] + 1] = hint
+  end
+
+  for lnum, _ in pairs(hints_by_lnum) do
+    if not next(new_hints_by_lnum[lnum]) then
+      hints_by_lnum[lnum] = nil
+    end
+  end
+
+  local countdown = function()
+    num_unprocessed = num_unprocessed - 1
+    if num_unprocessed == 0 then
+      bufstate.version = version
+      vim.cmd('redraw!')
+    end
+  end
+
+  local redraw = false
+  for lnum, new_line_hints in pairs(new_hints_by_lnum) do
+    local num_unresolved = 0
+    for _, hint in pairs(new_line_hints) do
+      if not hint._resolved then
+        num_unresolved = num_unresolved + 1
+      end
+    end
+    local current_line_hints = hints_by_lnum[lnum]
+    if not next(current_line_hints or {}) then
+      hints_by_lnum[lnum] = new_line_hints
+      redraw = true
+    end
+    for _, hint in pairs(new_line_hints) do
+      if hint._resolved then
+        countdown()
+      else
+        client.request('inlayHint/resolve', hint, function(err, result)
+          if err then
+            hint._resolved = false
+            return
+          end
+          if result then
+            hint.kind = result.kind and result.kind or hint.kind
+            hint.textEdits = result.textEdits and result.textEdits or hint.textEdits
+            hint.label = result.label and result.label or hint.label
+          end
+          hint._resolved = true
+          num_unresolved = num_unresolved - 1
+          if num_unresolved == 0 and current_line_hints then
+            hints_by_lnum[lnum] = new_line_hints
+          end
+          countdown()
+        end)
+      end
+    end
+  end
+  if redraw then
+    vim.cmd('redraw!')
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/inlayHint`
+---
+function M.on_inlayhint(err, result, ctx, _)
+  if err then
+    local _ = log.error() and log.error('inlayhint', err)
+    return
+  end
+  M.save(result, ctx.bufnr, ctx.client_id, ctx.version)
+end
+
+---@private
+local function resolve_bufnr(bufnr)
+  return bufnr == 0 and api.nvim_get_current_buf() or bufnr
+end
+
+--- Refresh the inlay hints for the current buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---
+--- Example:
+--- <pre>vim
+---   autocmd BufEnter,InsertLeave,BufWritePost <buffer> lua vim.lsp.inlay_hint.refresh()
+--- </pre>
+---
+function M.refresh(bufnr)
```

Comment:
Not sure about this, but maybe it would be worth to future proof this with a `opts` table with `{ bufnr = .. }` instead in case we want/need to add additional options. (E.g. if we were to let users set the range)